File: payments.php
<?php
header('Content-Type: application/json');
require_once __DIR__ . '/utils.php';

$method = $_SERVER['REQUEST_METHOD'];
$users = read_json_file(USERS_FILE);
if (!is_array($users)) $users = [];

$payments = read_json_file(PAYMENTS_FILE);
if (!is_array($payments)) $payments = [];

function get_user_by_username($users, $username) {
  foreach ($users as $user) {
    if (($user['username'] ?? '') === $username) {
      return $user;
    }
  }
  return null;
}

function ensure_group_access($users, $group, $username) {
  $user = get_user_by_username($users, $username);
  if (!$user || ($user['group'] ?? '') !== $group) {
    http_response_code(403);
    echo json_encode(["success" => false, "message" => "Accesso non consentito al gruppo richiesto"]);
    exit;
  }
  return $user;
}

switch ($method) {
  case 'GET':
    $group = trim($_GET['group'] ?? '');
    $role = $_GET['role'] ?? 'user';
    $username = trim($_GET['username'] ?? '');
    $date = trim($_GET['date'] ?? '');
    if ($date === '') {
      $date = date('Y-m-d');
    }

    if ($group === '') {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Gruppo richiesto mancante"]);
      exit;
    }

    if ($role !== 'admin') {
      if ($username === '') {
        http_response_code(400);
        echo json_encode(["success" => false, "message" => "Utente richiesto mancante"]);
        exit;
      }
      ensure_group_access($users, $group, $username);
    }

    $groupMembers = [];
    foreach ($users as $user) {
      if (($user['group'] ?? '') === $group) {
        $groupMembers[] = $user['username'];
      }
    }

    $counts = [];
    foreach ($groupMembers as $member) {
      $counts[$member] = 0;
    }

    $log = [];
    foreach ($payments as $paymentDate => $groups) {
      if (!is_array($groups)) continue;
      $payer = trim((string)($groups[$group] ?? ''));
      if ($payer === '') continue;
      $counts[$payer] = ($counts[$payer] ?? 0) + 1;
      $log[] = ['date' => $paymentDate, 'username' => $payer];
    }

    usort($log, function ($a, $b) {
      return strcmp($b['date'], $a['date']);
    });

    $history = [];
    foreach ($counts as $member => $count) {
      $history[] = ['username' => $member, 'count' => $count];
    }

    usort($history, function ($a, $b) {
      if ($a['count'] === $b['count']) {
        return strcmp($a['username'], $b['username']);
      }
      return $b['count'] <=> $a['count'];
    });

    $payerForDate = null;
    if (isset($payments[$date]) && is_array($payments[$date])) {
      $recorded = trim((string)($payments[$date][$group] ?? ''));
      if ($recorded !== '') {
        $payerForDate = ['username' => $recorded, 'date' => $date];
      }
    }

    echo json_encode([
      "success" => true,
      "group" => $group,
      "date" => $date,
      "payer" => $payerForDate,
      "totals" => $history,
      "log" => $log
    ], JSON_UNESCAPED_UNICODE);
    break;

  case 'POST':
    $raw = file_get_contents('php://input');
    $payload = json_decode($raw, true);
    if (!is_array($payload)) $payload = [];

    $group = trim($payload['group'] ?? '');
    $payer = trim($payload['payer'] ?? '');
    $date = trim($payload['date'] ?? '');
    $role = $payload['role'] ?? 'user';
    $actor = trim($payload['actor'] ?? '');

    if ($date === '') {
      $date = date('Y-m-d');
    }

    if ($group === '' || $payer === '') {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Dati mancanti o non validi per il pagamento"]);
      exit;
    }

    $payerUser = get_user_by_username($users, $payer);
    if (!$payerUser) {
      http_response_code(404);
      echo json_encode(["success" => false, "message" => "Utente non trovato"]);
      exit;
    }

    if (($payerUser['group'] ?? '') !== $group) {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "L'utente selezionato non appartiene al gruppo"]);
      exit;
    }

    if ($role !== 'admin') {
      if ($actor === '') {
        $actor = $payer;
      }
      ensure_group_access($users, $group, $actor);
      if ($actor !== $payer) {
        http_response_code(403);
        echo json_encode(["success" => false, "message" => "Non puoi registrare il pagamento per un altro utente"]);
        exit;
      }
    }

    if (!isset($payments[$date]) || !is_array($payments[$date])) {
      $payments[$date] = [];
    }
    $payments[$date][$group] = $payer;

    if (!write_json_file(PAYMENTS_FILE, $payments)) {
      http_response_code(500);
      echo json_encode(["success" => false, "message" => "Impossibile salvare il pagamento"]);
      exit;
    }

    echo json_encode(["success" => true]);
    break;

  default:
    http_response_code(405);
    echo json_encode(["success" => false, "message" => "Metodo non supportato"]);
}

File: app/routers/payments.py
from __future__ import annotations
from typing import Any, Dict, List, Optional
from datetime import date
from fastapi import APIRouter, Query
from ..storage import USERS_FILE, PAYMENTS_FILE, read_json_file, write_json_file
from ..common import success, error_response
from ..utils import ensure_group_access

router = APIRouter()

@router.get("/payments", response_model=None)
def get_payments(
    group: str = Query(""),
    role: str = Query("user"),
    username: str = Query(""),
    date_param: Optional[str] = Query(default=None, alias="date"),
):
    group = group.strip()
    username = username.strip()
    requested_date = date_param.strip() if isinstance(date_param, str) and date_param else date.today().isoformat()

    if not group:
        return error_response(400, "Gruppo richiesto mancante")

    users = read_json_file(USERS_FILE) or []
    payments = read_json_file(PAYMENTS_FILE) or {}
    if not isinstance(payments, dict):
        payments = {}

    if role != "admin":
        if not username:
            return error_response(400, "Utente richiesto mancante")
        if not ensure_group_access(users, group, username):
            return error_response(403, "Accesso non consentito al gruppo richiesto")

    group_members = [user.get("username") for user in users if user.get("group") == group]

    counts = {member: 0 for member in group_members}
    log: List[Dict[str, Any]] = []

    for payment_date, per_group in payments.items():
        if isinstance(per_group, dict):
            payer = str(per_group.get(group, "")).strip()
            if payer:
                counts[payer] = counts.get(payer, 0) + 1
                log.append({"date": payment_date, "username": payer})

    log.sort(key=lambda entry: entry["date"], reverse=True)

    history = [
        {"username": member, "count": counts.get(member, 0)}
        for member in sorted(counts.keys())
    ]
    history.sort(key=lambda entry: (-entry["count"], entry["username"]))

    payer_for_date = None
    day_record = payments.get(requested_date)
    if isinstance(day_record, dict):
        recorded = str(day_record.get(group, "")).strip()
        if recorded:
            payer_for_date = {"username": recorded, "date": requested_date}

    return success(group=group, date=requested_date, payer=payer_for_date, totals=history, log=log)


@router.post("/payments", response_model=None)
def register_payment(payload: Dict[str, Any]):
    group = str(payload.get("group", "")).strip()
    payer = str(payload.get("payer", "")).strip()
    role = payload.get("role", "user")
    actor = str(payload.get("actor", "")).strip()
    requested_date = str(payload.get("date", "")).strip() or date.today().isoformat()

    if not group or not payer:
        return error_response(400, "Dati mancanti o non validi per il pagamento")

    users = read_json_file(USERS_FILE) or []
    payer_user = next((user for user in users if user.get("username") == payer), None)
    if not payer_user:
        return error_response(404, "Utente non trovato")
    if payer_user.get("group") != group:
        return error_response(400, "L'utente selezionato non appartiene al gruppo")

    if role != "admin":
        actor = actor or payer
        if not ensure_group_access(users, group, actor):
            return error_response(403, "Accesso non consentito al gruppo richiesto")
        if actor != payer:
            return error_response(403, "Non puoi registrare il pagamento per un altro utente")

    payments = read_json_file(PAYMENTS_FILE) or {}
    if not isinstance(payments, dict):
        payments = {}

    day_record = payments.get(requested_date)
    if not isinstance(day_record, dict):
        day_record = {}
    day_record[group] = payer
    payments[requested_date] = day_record

    if not write_json_file(PAYMENTS_FILE, payments):
        return error_response(500, "Impossibile salvare il pagamento")

    return success()
